core/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala

```diff
@@ -290,18 +290,39 @@ object Checkpoints extends DeltaLogging {
     // log store and decide whether to use rename.
     val useRename = deltaLog.store.isPartialWriteVisible(deltaLog.logPath, hadoopConf)
 
-    val checkpointRowCount = spark.sparkContext.longAccumulator("checkpointRowCount")
-    val numOfFiles = spark.sparkContext.longAccumulator("numOfFiles")
+    val sessionConf = spark.sessionState.conf
+    val maxActionsPerFile = sessionConf.getConf(DeltaSQLConf.DELTA_CHECKPOINT_MAX_ACTIONS_PER_FILE)
+
+    val res = snapshot.stateDF
+      .groupBy()
+      .agg(count("*").alias("rowCount"), count("add").alias("numOfFiles"))
+      .collect()
+      .head
+
+    val checkpointRowCount = res.getLong(0)
+    val numOfFiles = res.getLong(1)
+
+    val numCheckpointFiles = maxActionsPerFile.map { maxActions =>
+      (checkpointRowCount - 1) / maxActions + 1
+    }.getOrElse(1L)
+
+    val checkpointPaths = if (numCheckpointFiles > 1) {
+      checkpointFileWithParts(snapshot.path, snapshot.version, numCheckpointFiles.toInt)
```

Comment:
just to be safe, add a check to `numCheckpointFiles.toInt` to throw exception if we exceeded the int limit.

Reply:
any specific exception I should throw?

Reply:
and looking at the scala api, is `isValidInt` a good way to test? or is just comparing to Int.MaxValue better?